osh/string_ops.py

```diff
@@ -61,36 +61,64 @@ def _CheckContinuationByte(byte):
     raise util.InvalidUtf8(INVALID_CONT)
 
 
+def _Utf8CharLen(starting_byte):
+  if (starting_byte >> 7) == 0b0:
+    return 1
+  elif (starting_byte >> 5) == 0b110:
+    return 2
+  elif (starting_byte >> 4) == 0b1110:
+    return 3
+  elif (starting_byte >> 3) == 0b11110:
+    return 4
+  else:
+    raise util.InvalidUtf8(INVALID_START)
+
+
 def _NextUtf8Char(s, i):
   """
-  Given a string and a byte offset, returns the byte position of the next char.
+  Given a string and a byte offset, returns the byte position after
+  the character at this position.  Usually this is the position of the
+  next character, but for the last character in the string, it's the
+  position just past the end of the string.
+
   Validates UTF-8.
   """
   byte_as_int = ord(s[i])  # Should never raise IndexError
 
   try:
-    if (byte_as_int >> 7) == 0b0:
-      i += 1
-    elif (byte_as_int >> 5) == 0b110:
-      _CheckContinuationByte(s[i+1])
-      i += 2
-    elif (byte_as_int >> 4) == 0b1110:
-      _CheckContinuationByte(s[i+1])
-      _CheckContinuationByte(s[i+2])
-      i += 3
-    elif (byte_as_int >> 3) == 0b11110:
-      _CheckContinuationByte(s[i+1])
-      _CheckContinuationByte(s[i+2])
-      _CheckContinuationByte(s[i+3])
-      i += 4
-    else:
-      raise util.InvalidUtf8(INVALID_START)
+    length = _Utf8CharLen(byte_as_int)
+    for j in xrange(i + 1, i + length):
+      _CheckContinuationByte(s[j])
+    i += length
   except IndexError:
     raise util.InvalidUtf8(INCOMPLETE_CHAR)
 
   return i
 
 
+def _PreviousUtf8Char(s, i):
+  """
+  Given a string and a byte offset, returns the position of the
```

Comment:
OK this looks like a short and useful piece of code.

1) Can you write unit tests for it in `osh/string_ops_test.py` ?  I don't think the error cases are covered by the spec tests.

2) Is there a URL to explain this algorithm?  I think I get it but it's not obvious upon reading.

Here's a link I found through Googling, but the explanation is unclear and some people are quibbling with it:

https://stackoverflow.com/questions/22257486/iterate-backwards-through-a-utf8-multibyte-string

Reply:
1. Sure, sorry, didn't even consciously notice the test module.  Still learning my way around the code base :)

2. I can try to find another one, but when I was working on it I only referred to that stackoverflow post and the description on wikipedia.  I'll try to come up with a clear explanation in the code.

Reply:
0b9445adea807545bb1bc8f49ea5424db78c0c26 -- added tests for this function, as well as _NextUtf8Char, which lacked them.
1e611ed4514fec8e0afd4fe8d480e91bbcea2063 -- added comment to explain algorithm in _PreviousUtf8Char